src/perf/lib/PerfServer.cpp

```diff
@@ -415,11 +559,134 @@ PerfServer::TcpSendCompleteCallback(
+    //
+    uint16_t ThreadFlags = CXPLAT_THREAD_FLAG_SET_IDEAL_PROC | CXPLAT_THREAD_FLAG_SET_AFFINITIZE;
+    CXPLAT_THREAD_CONFIG Config = { ThreadFlags, PartitionIndex, "DelayWorker", WorkerThread, this };
+    if (QUIC_FAILED(Thread.Create(&Config))) {
+        WriteOutput("CxPlatThreadCreate FAILED\n");
+        return false;
+    }
+
+    Initialized = true;
+    return true;
+}
+
+void DelayWorker::Shutdown()
+{
+    Shuttingdown = true;
+
+    if (Initialized) {
+        WakeWorkerThread();
+        Initialized = false;
+        //
+        // delete any pending work items
+        //
+        Lock.Acquire();
+        DelayedWorkContext* CurrentWorkItem = WorkItems;
+        while (nullptr != CurrentWorkItem) {
+            DelayedWorkContext* NextWorkItem;
+            if (WorkItemsTail == &CurrentWorkItem->Next) {
+                NextWorkItem = nullptr;
+            } else {
+                NextWorkItem = CurrentWorkItem->Next;
+            }
+            delete CurrentWorkItem;
+            CurrentWorkItem = NextWorkItem;
+        }
+        WorkItems = nullptr;
+        Lock.Release();
+    }
+}
+
+void DelayWorker::WakeWorkerThread() {
+    WakeEvent.Set();
+}
+
+CXPLAT_THREAD_CALLBACK(DelayWorker::WorkerThread, Context)
+{
+    DelayWorker* This = (DelayWorker*)Context;
+    while (DelayedWork(This)) {
+        This->WakeEvent.WaitForever();
+    }
+    CXPLAT_THREAD_RETURN(0);
+}
+
+BOOLEAN
+DelayWorker::DelayedWork(
+    _Inout_ void* Context
+    )
+{
+    DelayWorker* This = (DelayWorker*)Context;
+    DelayedWorkContext* WorkItem;
+    DelayedWorkContext* NextWorkItem;
+
+    do {
+        if (This->Shuttingdown) {
+            This->DoneEvent.Set();
+            return FALSE;
+        }
+
+        WorkItem = nullptr;
+        NextWorkItem = nullptr;
+
+        This->Lock.Acquire();
+        if (nullptr != This->WorkItems) {
+            WorkItem = This->WorkItems;
+            if (nullptr != WorkItem) {
+                NextWorkItem = This->WorkItems = WorkItem->Next;
+                if (This->WorkItemsTail == &WorkItem->Next) {
+                    This->WorkItemsTail = &This->WorkItems;
+                }
+                WorkItem->Next = nullptr;
+            }
+        }
+        This->Lock.Release();
+
+        if (nullptr != WorkItem) {
+            This->Server->SimulateDelay();
+            if (WorkItem->Context->IsActive()) {
+                This->Server->SendResponse(WorkItem->Context, WorkItem->Handle, WorkItem->IsTcp);
+            }
+            WorkItem->Context->Release();
+            delete WorkItem;
+        }
+    } while (nullptr != NextWorkItem);
+
+    return TRUE;
+}
+
+void
+DelayWorker::QueueWork(
+    _In_ StreamContext* Context,
+    _In_ void* Handle,
+    _In_ bool IsTcp
+    )
+{
+    DelayedWorkContext* Work = new (std::nothrow) DelayedWorkContext();
+
+    Work->Context = Context;
+    Work->Handle = Handle;
+    Work->IsTcp = IsTcp;
```

Comment:
The handle and IsTcp flag are already saved on the context, so no need to save it on the Work item.